src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/groupms2/GroupMS2Task.java

```
@@ -0,0 +1,164 @@
+
+    // Initialize.
+    this.project = project;
+    parameters = parameterSet;
+    rtTol = parameters.getParameter(GroupMS2Parameters.rtTol).getValue();
+    mzTol = parameters.getParameter(GroupMS2Parameters.mzTol).getValue();
+    limitRTByFeature = parameters.getParameter(GroupMS2Parameters.limitRTByFeature).getValue();
+    this.list = list;
+    processedRows = 0;
+    totalRows = 0;
+  }
+
+  @Override
+  public double getFinishedPercentage() {
+
+    return totalRows == 0 ? 0.0 : (double) processedRows / (double) totalRows;
+  }
+
+  @Override
+  public String getTaskDescription() {
+
+    return "Adding all MS2 scans to their features in list " + list.getName();
+  }
+
+  @Override
+  public void run() {
+
+    try {
+      setStatus(TaskStatus.PROCESSING);
+
+      totalRows = list.getNumberOfRows();
+      // for all features
+      for (PeakListRow row : list.getRows()) {
+        for (Feature f : row.getPeaks()) {
+          if (getStatus() == TaskStatus.ERROR)
+            return;
+          if (isCanceled())
+            return;
+
+          RawDataFile raw = f.getDataFile();
+          IntArrayList scans = new IntArrayList();
+          int best = f.getRepresentativeScanNumber();
+          double frt = f.getRT();
+          double fmz = f.getMZ();
+          Range<Double> rtRange = f.getRawDataPointsRTRange();
+          int i = best;
+          // left
+          while (i > 1) {
+            try {
+              i--;
+              Scan scan = raw.getScan(i);
+              if ((!limitRTByFeature || rtRange.contains(scan.getRetentionTime()))
+                  && rtTol.checkWithinTolerance(frt, scan.getRetentionTime())) {
+                if (scan.getPrecursorMZ() != 0
+                    && mzTol.checkWithinTolerance(fmz, scan.getPrecursorMZ()))
+                  scans.add(i);
+              } else {
+                // end of loop - out of tolerance
+                break;
+              }
+            } catch (Exception e) {

```

At least add e.printStackTrace(), you should never swallow exceptions without warning.

---

Yeah changed it to a null checking. 
The RawDataFile.getScan(i) method should be @Nullable 
The user can enter any index which is out of range or just not in the HashTable (the scan numbers do not have to be contiguous).